File: PuzzleBot_Development/traffic_light/traffic_light/detect.py
import numpy as np
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from std_msgs.msg import Bool, String

from traffic_light.TraficLightDetection import TrafficLightDetection


def imgmsg_to_bgr(msg: Image) -> np.ndarray:
    """Convert a sensor_msgs/Image to a BGR numpy array without cv_bridge."""
    raw = np.frombuffer(msg.data, dtype=np.uint8)
    if msg.encoding in ('bgr8', 'rgb8'):
        img = raw.reshape((msg.height, msg.width, 3))
        if msg.encoding == 'rgb8':
            img = img[:, :, ::-1]  # RGB → BGR
    elif msg.encoding == 'mono8':
        img = raw.reshape((msg.height, msg.width))
    else:
        # fallback: treat as bgr8
        img = raw.reshape((msg.height, msg.width, -1))
    return np.ascontiguousarray(img)


def bgr_to_imgmsg(bgr: np.ndarray, stamp=None,
                  frame_id: str = 'traffic_light_debug') -> Image:
    """Pack a BGR numpy array into a sensor_msgs/Image (no cv_bridge)."""
    bgr = np.ascontiguousarray(bgr)
    msg = Image()
    msg.height = int(bgr.shape[0])
    msg.width = int(bgr.shape[1])
    msg.encoding = 'bgr8'
    msg.is_bigendian = 0
    msg.step = int(bgr.shape[1] * 3)
    msg.data = bgr.tobytes()
    msg.header.frame_id = frame_id
    if stamp is not None:
        msg.header.stamp = stamp
    return msg


class TrafficLightNode(Node):
    # States that allow the controller to move. red and yellow stop it.
    # 'none' (no light detected) also allows movement so detection dropouts
    # don't bring the robot to a halt.
    GO_STATES = {'green', 'none'}

    GO_TOPIC = '/traffic_light/go'
    STATE_TOPIC = '/traffic_light/state'
    DEBUG_IMAGE_TOPIC = '/traffic_light/debug_image'

    def __init__(self):
        super().__init__('traffic_light_detector')
        self.declare_parameter('debug', False)
        self.declare_parameter('confirm_frames', 3)
        self.declare_parameter('go_publish_hz', 5.0)
        debug = self.get_parameter('debug').get_parameter_value().bool_value
        self.confirm_frames = max(
            1, int(self.get_parameter('confirm_frames').value)
        )
        go_hz = max(0.1, float(self.get_parameter('go_publish_hz').value))
        self.detector = TrafficLightDetection(debug=debug)

        self.subscription = self.create_subscription(
            Image,
            '/video_source/raw',
            self._image_callback,
            10,
        )
        self.state_pub = self.create_publisher(String, self.STATE_TOPIC, 10)
        self.go_pub = self.create_publisher(Bool, self.GO_TOPIC, 10)
        # Live per-frame debug visualization — NOT gated by the confirmation
        # filter; publishes whatever the detector saw this frame.
        self.debug_image_pub = self.create_publisher(
            Image, self.DEBUG_IMAGE_TOPIC, 5
        )

        # Debounce: only update the published "go" once the same state has
        # repeated for confirm_frames in a row.
        self._candidate_state = None
        self._candidate_count = 0
        self._confirmed_state = None
        # Default to True ("go") so the receiver gets a sane heartbeat from
        # the very first tick, before any detection has happened.
        self._current_go = True

        # Steady heartbeat on /traffic_light/go — the receiver gets True or
        # False at go_publish_hz regardless of camera framerate or whether
        # a new confirmation has fired.
        self.go_timer = self.create_timer(1.0 / go_hz, self._publish_go)

        self.get_logger().info(
            f'Traffic light detector started '
            f'(confirm_frames={self.confirm_frames}, '
            f'go_publish_hz={go_hz:g}).'
        )

    def _image_callback(self, msg: Image):
        frame = imgmsg_to_bgr(msg)
        raw_state = self.detector.detect_state(frame)

        # Always publish the raw per-frame state for downstream debugging.
        self.state_pub.publish(String(data=raw_state))

        # Live debug visualization — every frame, no debounce.
        dbg = self.detector.last_debug_frame
        if dbg is not None:
            self.debug_image_pub.publish(
                bgr_to_imgmsg(dbg, stamp=msg.header.stamp)
            )

        confirmed = self._debounce(raw_state)
        if confirmed is not None:
            self._on_confirmed_state(confirmed)

    def _debounce(self, raw_state: str):
        """
        Return the new *confirmed* state once raw_state has repeated for
        `confirm_frames` consecutive frames and differs from the previous
        confirmed state. Otherwise return None.
        """
        if raw_state == self._candidate_state:
            self._candidate_count += 1
        else:
            self._candidate_state = raw_state
            self._candidate_count = 1

        if (self._candidate_count >= self.confirm_frames
                and raw_state != self._confirmed_state):
            self._confirmed_state = raw_state
            return raw_state
        return None

    def _on_confirmed_state(self, state: str):
        """Update the latched go value and log the transition.

        Publishing is left to the timer (`_publish_go`) so the receiver
        always sees a steady stream at `go_publish_hz`.
        """
        go = state in self.GO_STATES
        self._current_go = go
        self.get_logger().info(f'detected: {state}  -> go={go}')

    def _publish_go(self):
        """Timer callback — publish the latest go value on every tick."""
        self.go_pub.publish(Bool(data=self._current_go))


def main(args=None):
    rclpy.init(args=args)
    node = TrafficLightNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: PuzzleBot_Development/traffic_light/src/detect.cpp
#include <chrono>
#include <cstring>
#include <memory>
#include <optional>
#include <string>

#include <opencv2/opencv.hpp>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "std_msgs/msg/bool.hpp"
#include "std_msgs/msg/string.hpp"

#include "traffic_light_detection.hpp"

namespace
{

cv::Mat imgmsg_to_bgr(const sensor_msgs::msg::Image & msg)
{
  const int rows = static_cast<int>(msg.height);
  const int cols = static_cast<int>(msg.width);
  if (rows <= 0 || cols <= 0 || msg.data.empty()) {
    return cv::Mat();
  }
  uint8_t * data = const_cast<uint8_t *>(msg.data.data());

  if (msg.encoding == "bgr8") {
    return cv::Mat(rows, cols, CV_8UC3, data).clone();
  }
  if (msg.encoding == "rgb8") {
    cv::Mat src(rows, cols, CV_8UC3, data);
    cv::Mat bgr;
    cv::cvtColor(src, bgr, cv::COLOR_RGB2BGR);
    return bgr;
  }
  if (msg.encoding == "mono8") {
    cv::Mat gray(rows, cols, CV_8UC1, data);
    cv::Mat bgr;
    cv::cvtColor(gray, bgr, cv::COLOR_GRAY2BGR);
    return bgr;
  }
  // Fallback: assume bgr8.
  return cv::Mat(rows, cols, CV_8UC3, data).clone();
}

sensor_msgs::msg::Image bgr_to_imgmsg(
  const cv::Mat & bgr, const builtin_interfaces::msg::Time & stamp,
  const std::string & frame_id)
{
  sensor_msgs::msg::Image msg;
  msg.height = static_cast<uint32_t>(bgr.rows);
  msg.width = static_cast<uint32_t>(bgr.cols);
  msg.encoding = "bgr8";
  msg.is_bigendian = 0;
  msg.step = static_cast<uint32_t>(bgr.cols * 3);
  msg.data.resize(static_cast<size_t>(msg.step) * msg.height);
  if (bgr.isContinuous()) {
    std::memcpy(msg.data.data(), bgr.data, msg.data.size());
  } else {
    for (int r = 0; r < bgr.rows; ++r) {
      std::memcpy(
        msg.data.data() + static_cast<size_t>(r) * msg.step,
        bgr.ptr(r), msg.step);
    }
  }
  msg.header.stamp = stamp;
  msg.header.frame_id = frame_id;
  return msg;
}

}  // namespace

class TrafficLightNode : public rclcpp::Node
{
public:
  TrafficLightNode()
  : Node("traffic_light_detector")
  {
    debug_ = this->declare_parameter<bool>("debug", false);
    confirm_frames_ = std::max(
      1, static_cast<int>(this->declare_parameter<int>("confirm_frames", 3)));
    const double go_hz = std::max(
      0.1, this->declare_parameter<double>("go_publish_hz", 5.0));

    detector_ = std::make_unique<traffic_light::TrafficLightDetector>(debug_);

    state_pub_ = this->create_publisher<std_msgs::msg::String>(
      "/traffic_light/state", 10);
    go_pub_ = this->create_publisher<std_msgs::msg::Bool>(
      "/traffic_light/go", 10);
    debug_image_pub_ = this->create_publisher<sensor_msgs::msg::Image>(
      "/traffic_light/debug_image", 5);

    // KeepLast(1): always process the freshest camera frame and drop stale
    // ones instead of working through a 10-frame backlog (which made
    // detection lag noticeable on slower hardware).
    image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
      "/video_source/raw", rclcpp::QoS(1),
      std::bind(&TrafficLightNode::image_callback, this,
        std::placeholders::_1));

    go_timer_ = this->create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::duration<double>(1.0 / go_hz)),
      std::bind(&TrafficLightNode::publish_go, this));

    RCLCPP_INFO(
      this->get_logger(),
      "Traffic light detector started (confirm_frames=%d, go_publish_hz=%g).",
      confirm_frames_, go_hz);
  }

private:
  void image_callback(sensor_msgs::msg::Image::ConstSharedPtr msg)
  {
    cv::Mat frame = imgmsg_to_bgr(*msg);
    if (frame.empty()) {
      return;
    }

    // Skip building the debug composite when nobody's watching — debug
    // rendering (clone + 6 tiles + hconcat/vconcat + putText) was a big
    // chunk of the per-frame cost in the Python version.
    const bool want_debug =
      debug_ || debug_image_pub_->get_subscription_count() > 0;

    const traffic_light::State raw =
      detector_->detect_state(frame, want_debug);
    const char * raw_name = traffic_light::to_string(raw);

    std_msgs::msg::String state_msg;
    state_msg.data = raw_name;
    state_pub_->publish(state_msg);

    if (want_debug && !detector_->last_debug_frame().empty()) {
      debug_image_pub_->publish(
        bgr_to_imgmsg(
          detector_->last_debug_frame(), msg->header.stamp,
          "traffic_light_debug"));
    }

    const auto confirmed = debounce(raw);
    if (confirmed.has_value()) {
      on_confirmed_state(*confirmed);
    }
  }

  std::optional<traffic_light::State> debounce(traffic_light::State raw)
  {
    if (have_candidate_ && raw == candidate_state_) {
      ++candidate_count_;
    } else {
      candidate_state_ = raw;
      candidate_count_ = 1;
      have_candidate_ = true;
    }
    if (candidate_count_ >= confirm_frames_ &&
      (!have_confirmed_ || raw != confirmed_state_))
    {
      confirmed_state_ = raw;
      have_confirmed_ = true;
      return raw;
    }
    return std::nullopt;
  }

  void on_confirmed_state(traffic_light::State s)
  {
    const bool go =
      s == traffic_light::State::Green || s == traffic_light::State::None;
    current_go_ = go;
    RCLCPP_INFO(
      this->get_logger(), "detected: %s  -> go=%s",
      traffic_light::to_string(s), go ? "true" : "false");
  }

  void publish_go()
  {
    std_msgs::msg::Bool b;
    b.data = current_go_;
    go_pub_->publish(b);
  }

  bool debug_;
  int confirm_frames_;
  std::unique_ptr<traffic_light::TrafficLightDetector> detector_;

  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr state_pub_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr go_pub_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr debug_image_pub_;
  rclcpp::TimerBase::SharedPtr go_timer_;

  traffic_light::State candidate_state_ = traffic_light::State::None;
  int candidate_count_ = 0;
  bool have_candidate_ = false;
  traffic_light::State confirmed_state_ = traffic_light::State::None;
  bool have_confirmed_ = false;
  bool current_go_ = true;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TrafficLightNode>());
  rclcpp::shutdown();
  return 0;
}
